Web Animations API (#2956)
GoogleChrome/web.dev: src/site/content/en/blog/web-animations/index.md

---

> title: The Web Animations API just hit Chrome stable — here's what you need to know

Reviewer: Please make your title more like the [our other ones](https://web.dev/blog/).

Always avoid the terms "stable", "beta", and "canary". Articles will be on the web for years. These terms are only accurate for six weeks.

> subhead: Wrangling your web animations is about to get much easier

Author: I changed the title to "What's coming to Chrome 84 in the Web Animations API" -- would you prefer this or "Wrangling your web animations is about to get much easier"

> subhead: Wrangling your web animations is about to get much easier

Reviewer: Yes I would. This isn't just a personal prefernce. This change would make this more in line with preferred practice.

> subhead: Wrangling your web animations is about to get much easier

Reviewer 2: See my suggestion below

authors:
  - una
  - kevinellis
date: 2020-05-27
hero: hero.png
# You can adjust the position of your hero image with this property.
# Values: top | bottom | center (default)
# hero_position: bottom
# thumbnail: thumbnail.png
description: Wrangling your web animations is about to get much easier

> tags:

Reviewer 2: Here's the list of "canonical" tags: https://github.com/GoogleChrome/web.dev/blob/master/src/site/_data/postTags.js

Using these tags causes the nice little category bubbles to show up at the bottom of the post (e.g. `CSS` and `Fonts` in the image below):

![image](https://user-images.githubusercontent.com/4713486/82955425-ab230880-9f63-11ea-8264-881bc0509f2b.png)

---

> title: "What's coming to Chrome 84 in the Web Animations API"
> subhead: Wrangling your web animations is about to get much easier

Reviewer 2: ```suggestion
title: Web Animations API improvements in Chromium 84
subhead: Orchestrating animations with promises, performance improvements with replaceable animations, smoother animations with composite modes, and more.
```

* Presumably these changes will also land in Microsoft Edge, in which case we should use `Chromium` throughout the article rather than `Chrome`. My proposed rephrasing of the title also provides more of a hint that there is a lot of new goodness coming (and it's more concise).
* The current `subhead` doesn't give me any hint as to what has changed. I want the subtitle to quickly inform users about what has changed so that they can decide whether to invest more time by reading the article.

Author: 👍

<figure class="w-figure">
  <img class="w-screenshot" src='./waapi-timeline.png' alt="The Web Animations API first hit Chrome in version 36, July of 2014. Now the spec is going to be complete, in version 84, launching July 2020.">
  <figcaption class="w-figcaption">
    The long history of the Web Animations API.
  </figcaption>
</figure>

The [Web Animations API](https://www.w3.org/TR/web-animations-1/) is a tool that enables developers to write [imperative](https://www.youtube.com/watch?v=WaNoqBAp8NI) animations with JavaScript. It was written to underpin both CSS animation and transition implementations and enable future effects to be developed, as well as existing effects to be composed and timed.

While it has a long history, originally launching in Chrome 36, the latest release, Chrome 84, brings us a slew of previously unsupported features. 

## Getting started

> Creating an animation via the web animations API should feel very familiar if you've used `"@keyframe"` rules. First you'll need to create a Keyframe Object. What might look like [this](https://codepen.io/una/pen/RwWMvPw) in CSS:

Reviewer: Just to clarify, my request for quotes around keywords was only for the names of the modes because the quotes are part of what needs to be in code. If I led you to overcorrect, I apologize.

+1 on the link. Try to link on text that describes what's at the end of the link.

  { transform: 'scale(0)' }, 
  { transform: 'scale(1)' }, 
];
```

Where you set parameters for animation in CSS:

```css
.modal {
  animation: openAnimation 1s 1 ease-in;
}
```

You would set in JS:

```js
document.querySelector('.modal').animate(
    openAnimation ,  {
    duration: 1000, // 1s
    iterations: 1, // single iteration
    easing: 'ease-in' // easing function
  }
);
```

The amount of code is about the same, but with JavaScript, you get a couple of superpowers that you don't have with CSS alone. This includes the ability to sequence effects, and an increased control of their play states.

{% Aside %}
  Hyphenated property names become camel case when used in keyframes (i.e. `"background-color"` to `"backgroundColor"`)
{% endAside %}

## Orchestrating animations with promises

In Chrome 84, you now have two methods that can be used with promises: `"animation.ready"` and `"animation.finished"`.

- `animation.ready` enables you to wait for pending changes to take effect (i.e. switching between playback control methods such as play and pause).
- `animation.finished` provides a means of executing custom JavaScript code when an animation is complete.

> Let's continue with our example, and create an orchestrated animation chain with `"animation.finished"`. Here, you have a vertical transformation (`"scaleY"`), followed by a horizontal transformation (`"scaleX"`), followed by an opacity change on a child element: 

Reviewer: "Let's continue with our example" to "I'll continue with the example"

<figure class="w-figure">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/web-animations/modal-open.mp4">
  </video>
  <figcaption class="w-figcaption">
    Applying transformations and opacity to an opening modal element. <a href="https://codepen.io/una/pen/dyYKJMz">See Demo on Codepen</a>
  </figcaption>

> </figure>

Reviewer 2: Please embed this as a Glitch onto the page: https://web.dev/handbook/web-dev-components/#glitches 

[Example](https://web.dev/read-files/#read-content)

This provides a better experience because it enables readers to browse the code without leaving our site.

```js/1
const transformAnimation = modal.animate(openModal, openModalSettings);
transformAnimation.finished.then(() => { text.animate(fadeIn, fadeInSettings)});
```

> We've chained these animations using `"animation.finished.then()"` prior to executing the next animation set in the chain. This way, the animations appear in order, and you are even applying effects to different target elements with different options set (such as speed and ease).

Reviewer 2: ```suggestion
We've chained these animations using `animation.finished.then()` prior to executing the next animation set in the chain. This way, the animations appear in order, and you are even applying effects to different target elements with different options set (such as speed and ease).
```

> Within CSS, this would be cumbersome to recreate, especially when applying unique, yet sequenced animations to multiple elements. You'd have to use a `"@keyframe"`, sort out the correct timing percentages to place the animations, and use `"animation-delay"` prior to triggering the animations in the sequence.

Reviewer 2: ```suggestion
Within CSS, this would be cumbersome to recreate, especially when applying unique, yet sequenced animations to multiple elements. You'd have to use a `@keyframe`, sort out the correct timing percentages to place the animations, and use `animation-delay` prior to triggering the animations in the sequence.
```

> ## Play, pause, and reverse

Reviewer 2: ```suggestion
### Example: Play, pause, and reverse
```

This section confused me. It's formatted as an H2, which implies that it's a new idea. I think it's actually just a continuation of the last section (*Orchestrating animations with promises*) so it should be formatted as an H3 instead. 

It was also confusing because there doesn't seem to be any new features here. That's why I'm suggesting adding `Example` to the section heading.

> What can open, should close! Luckily, since [Chrome 39](https://developers.google.com/web/updates/2014/12/web-animation-playback), the WAAPI has provided us the ability to play, pause, and reverse our animations.

Reviewer 2: ```suggestion
What can open, should close! Luckily, since [Chromium 39](https://developers.google.com/web/updates/2014/12/web-animation-playback), the Web Animations API has provided us the ability to play, pause, and reverse our animations.
```
We haven't defined the acronym `WAAPI` anywhere else in this article and it looks weird to me so I think it'd be safer to just use the full name throughout

> You can take the above animation, and give it a smooth, reversed animation when clicking the button again using `".reverse()"`. This way, you can create a smoother and more contextual interaction for our modal.

Reviewer 2: ```suggestion
You can take the above animation, and give it a smooth, reversed animation when clicking the button again using `.reverse()`. This way, you can create a smoother and more contextual interaction for our modal.
```

<figure class="w-figure">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/web-animations/modal-reverse.mp4">
  </video>
  <figcaption class="w-figcaption">
    An example of a modal opening and closing upon button click. <a href="https://codepen.io/kevers-google/details/LYpvdNG">See Demo on Codepen </a>
  </figcaption>
</figure>

> What you can do is create two play-pending animations (`"openModal"`, and an inline opacity transformation), and then pause one of the animations, delaying it until the other is finished. You can then use promises to wait for each to be finished before playing. Finally, you can check to see if a flag is set, and then reverse each animation.

Reviewer 2: ```suggestion
What you can do is create two play-pending animations (`openModal`, and an inline opacity transformation), and then pause one of the animations, delaying it until the other is finished. You can then use promises to wait for each to be finished before playing. Finally, you can check to see if a flag is set, and then reverse each animation.
```

```js/17-20
document.querySelector('button').addEventListener('click', () => {                                 
  // Create two play-pending animations.
  let animations = [
    document.querySelector('.modal').animate(openModal, openModalSettings),
    document.querySelector('.text').animate({ opacity: [0, 1] }, openModalSettings)
  ];

  // Pause one of the animations and delay it until the other is finished.
  if (open) {
    animations.forEach((anim) => {
      anim.currentTime = anim.effect.getTiming().duration;
      anim.reverse();
    });
    animations = animations.reverse();
  }
  
  // Run animations sequentially.
  animations[1].pause();  
  animations[0].finished.then(() => {
    animations[1].play();
  });
  open = !open;
});
```

> ## Dynamic interactions with partial keyframes

Reviewer 2: ```suggestion
### Example: Dynamic interactions with partial keyframes
```

<figure class="w-figure">
<video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/web-animations/retargetting.mp4">
  </video>  <figcaption class="w-figcaption">
    Retargeting example, where a mouse click adjusts the animation to a new location. <a href="https://codepen.io/una/details/BaoveQJ">See Demo on Codepen </a>
  </figcaption>

> </figure>

Reviewer 2: Please embed as a Glitch rather than linking to Codepen: https://web.dev/handbook/web-dev-components/#glitches

In this example, there is no specified start position. This is an example of using **partial keyframes**. The mouse handler does a few things here: it sets a new end location and triggers a new animation. The new start position is inferred from the current underlying position. 

New transitions can be triggered while existing ones are still running. This means that the current transition is interrupted, and a new one is created.

## Performance improvements with replaceable animations

> When creating animations based on events, such as on `"mousemove"`, a new animation is created each time, which can quickly consume memory and degrade performance.  To address this problem, replaceable animations were introduced, enabling automated cleanup, where finished animations are flagged as replaceable and automatically removed if replaced by another finished animation. Consider the following example:

Reviewer 2: ```suggestion
When creating animations based on events, such as on `'mousemove'`, a new animation is created each time, which can quickly consume memory and degrade performance.  To address this problem, replaceable animations were introduced in Chromium 83, enabling automated cleanup, where finished animations are flagged as replaceable and automatically removed if replaced by another finished animation. Consider the following example:
```
* Using single quotes because that's what the code sample uses
* Adding `in Chromium 84` to make it crystal clear that this is a new feature in v84

<figure class="w-figure">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/web-animations/comet-trail.mp4">
  </video>
  <figcaption class="w-figcaption">
  A comet trail animates when the mouse moves. <a href="https://codepen.io/kevers-google/details/gOaqXoQ">View Demo on Codepen</a>
  </figcaption>

> </figure>

Reviewer 2: Please embed as a Glitch rather than linking to Codepen

 ```js
elem.addEventListener('mousemove', evt => {
  rectangle.animate(
    { transform: translate(${evt.clientX}px, ${evt.clientY}px) },
    { duration: 500, fill: 'forwards' }
  );
});
```

Each time the mouse moves, the browser re-calculates the position for each ball in the comet trail and creates an animation to this new point. The browser now knows to remove old animations (enabling replacement) when:

> 1. The animation is finished
> 2. There is one or more animations higher in composite ordering that are also finished
> 3. The new animations are animating the same properties.

Reviewer 2: ```suggestion
1. The animation is finished.
2. There is one or more animations higher in composite ordering that are also finished.
3. The new animations are animating the same properties.
```
Consistent use of periods

> You can see exactly how many animations are being replaced by tallying up a counter with each removed animation, using `"anim.onremove"` to trigger the counter.

Reviewer 2: ```suggestion
You can see exactly how many animations are being replaced by tallying up a counter with each removed animation, using `anim.onremove` to trigger the counter.
```

There are a few additional methods to take your animation control even further:

> - `animation.onremove()` calls some custom JavaScript code when an animation is replaced.

Reviewer 2: ```suggestion
```
Isn't this the same method that we just mentioned in the previous paragraph?

- `animation.replaceState()` provides a means of tracking whether an animation is active, persisted, or removed.
- `animation.commitStyles()` updates the style of an element based on the underlying style along with all animations on the element in the composite order.
- `animation.persist()` marks an animation as non-replaceable.

{% Aside %}

>   The last two methods are commonly used with compositing modes, such as "add". Check out the demo below to see them in action.

Reviewer 2: ```suggestion
  `animation.commitStyles()` and `animation.persist()` are commonly used with compositing modes, such as "add". Check out the composite modes demo below to see them in action.
```
* Mentioning the methods explicitly will futureproof the article in the event that we update the list but forget to update this aside (it happens all the time)
* Making it clearer what "demo below" we're referring to

{% endAside %}

## Smoother animations with composite modes

> With the Web Animations API, you can now set the composite mode of your animations, meaning they can be additive or accumulative, in addition to the default mode of "replace". [Composite modes](https://css-tricks.com/additive-animation-web-animations-api/) allow developers to write distinct animations and have control over how effects are combined. Three composite modes are now supported: `"replace"` (the default mode), `"add"`, and `"accumulate"`.

Reviewer 2: ```suggestion
With the Web Animations API, you can now set the composite mode of your animations, meaning they can be additive or accumulative, in addition to the default mode of "replace". [Composite modes](https://css-tricks.com/additive-animation-web-animations-api/) allow developers to write distinct animations and have control over how effects are combined. Three composite modes are now supported: `'replace'` (the default mode), `'add'`, and `'accumulate'`.
```
Formatting should match the code samples, the code samples use single quotes

When you composite animations, a developer can write short, distinct effects and see them combined together. In the example below, we are applying a rotation and scale keyframe to each box, with the only adjustment being the composite mode, added as an option:

<figure class="w-figure">
<video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/web-animations/replace-add-accumulate.mp4">
  </video>  <figcaption class="w-figcaption">
    <a href="https://codepen.io/kevers-google/pen/wvKbaGM"> See Demo on Codepen </a>
  </figcaption>
</figure>

> In the default, `"replace"` composite mode, the final animation replaces the transform property and ends up at `rotate(360deg) scale(1.4)`. For `"add"`, composite adds the rotation and multiplies the scale, resulting in a final state of `rotate(720deg) scale(1.96)`. `"accumulate"` combines the transformations, resulting in `rotate(720deg) scale(1.8)`. For more on the intricacies of these composite modes, check out the [Web Animations Spec](https://www.w3.org/TR/web-animations-1/#the-compositeoperation-enumeration).

Reviewer 2: ```suggestion
In the default `'replace'` composite mode, the final animation replaces the transform property and ends up at `rotate(360deg) scale(1.4)`. For `'add'`, composite adds the rotation and multiplies the scale, resulting in a final state of `rotate(720deg) scale(1.96)`. `'accumulate'` combines the transformations, resulting in `rotate(720deg) scale(1.8)`. For more on the intricacies of these composite modes, check out [The CompositeOperation and CompositeOperationOrAuto enumerations](https://www.w3.org/TR/web-animations-1/#the-compositeoperation-enumeration) from the Web Animations spec.
```
* Fixes quotes
* Updates link text to match the name of the section that we're linking to

Let's take a look at a UI element example:

<figure class="w-figure">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/web-animations/dropdown.mp4">
  </video>
  <figcaption class="w-figcaption">
  A bouncy dropdown menu which has two composited animations applied to it. <a href="https://codepen.io/kevers-google/details/ExVrbEm">View Demo on Codepen</a>
  </figcaption>

> </figure>

Reviewer 2: Please embed as Glitch

Author: None of these will work on the current version of Chrome, so I kept them as images with Codepens they link to.  Do you think I should replace all the images with Glitch or have Glitch embeds in addition? Or just use Glitch over Codepen?

Reviewer 2: Videos/Images + Glitch embeds is ideal. If the reader just wants to get the main idea, they look at the video/image. If they want to dig in, they have the source code right there as a Glitch, without needing to leave the page (which reduces the odds that they get distracted and never come back).

> Here, two `"top"` animations are composited. The first is a macro-animation, which moves the dropdown by the full height of the menu itself as a slide-in effect from the top of the page, and the second, a micro-animation, applies a little bounce as it hits the bottom. Using the `"add"` composite mode enables a smoother transition.

Reviewer 2: ```suggestion
Here, two `top` animations are composited. The first is a macro-animation, which moves the dropdown by the full height of the menu itself as a slide-in effect from the top of the page, and the second, a micro-animation, applies a little bounce as it hits the bottom. Using the `'add'` composite mode enables a smoother transition.
```

```js/12
const dropDown = menu.animate(
    [
      { top: `${-menuHeight}px`, easing: 'ease-in' },
      { top: 0 }
    ], { duration: 300, fill: 'forwards' });

  dropDown.finished.then(() => {
    const bounce = menu.animate(
      [
        { top: '0px', easing: 'ease-in' },
        { top: '10px', easing: 'ease-out' },
        { ... }
      ], { duration: 300, composite: 'add' });
  });
```

> ### What's next for the web animations API

Reviewer 2: ```suggestion
### What's next for the Web Animations API
```
The rest of the article capitalizes the API name

---
title: Web Animations API improvements in Chromium 84
subhead: Orchestrating animations with promises, performance improvements with replaceable animations, smoother animations with composite modes, and more.
authors:
  - una
  - kevinellis
date: 2020-05-27

> hero: hero.png

Reviewer 2: I can't remember if we checked for this before, but can you try **lightly** optimizing this hero image with Squoosh? We have an image CDN so you shouldn't over-optimize it (or else it'll show up grainy for readers). The optimization is to keep the size of the repository down.

While it has a long history, originally launching in Chrome 36, the latest release, Chrome 84, brings us a slew of previously unsupported features. 

<figure class="w-figure">
  <img class="w-screenshot" src='./waapi-timeline.png' alt="The Web Animations API first hit Chrome in version 36, July of 2014. Now the spec is going to be complete, in version 84, launching July 2020.">
  <figcaption class="w-figcaption">
    The long history of the Web Animations API.
  </figcaption>
</figure>

## Getting started

Creating an animation via the web animations API should feel very familiar if you've used `"@keyframe"` rules. First you'll need to create a Keyframe Object. What might look like [this](https://codepen.io/una/pen/RwWMvPw) in CSS:

```css
@keyframes openAnimation {
  0% {
    transform: scale(0);
  }
  100% {
    transform: scale(1);
  }
}
```

Would look like [this](https://codepen.io/una/pen/abvYXJX) in JavaScript:

```js
const openAnimation = [
  { transform: 'scale(0)' }, 
  { transform: 'scale(1)' }, 
];
```

Where you set parameters for animation in CSS:

```css
.modal {
  animation: openAnimation 1s 1 ease-in;
}
```

> You would set in JS:

Reviewer: ```suggestion
you would set in JS:
```

```js
document.querySelector('.modal').animate(
    openAnimation, {
      duration: 1000, // 1s
      iterations: 1, // single iteration
      easing: 'ease-in' // easing function
    }
);
```

The amount of code is about the same, but with JavaScript, you get a couple of superpowers that you don't have with CSS alone. This includes the ability to sequence effects, and an increased control of their play states.

{% Aside %}

>   Hyphenated property names become camel case when used in keyframes (e.g. `"background-color"` to `"backgroundColor"`)

Reviewer 2: ```suggestion
  Hyphenated property names become camel case when used in keyframes (e.g. `background-color` to `backgroundColor`)
```

{% endAside %}

### Getting and setting animations beyond `element.animate()`

However, with the update, the Web Animations API is no longer restricted to animations created via `element.animate()`. We can manipulate CSS animations and transitions as well.

`getAnimations()` is a method that returns all animations on an element regardless of whether it was created via `element.animate()` or via CSS rules (CSS animation or transition). Here is an example of what this looks like:

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/waapi-getanimations?path=style.css&previewSize=100"
    title="waapi-getanimations on Glitch"

>     allow="geolocation; microphone; camera; midi; vr; encrypted-media"

Reviewer 2: ```suggestion
```

It looks like you used the embed code from Glitch (rather than [our code in the handbook](https://web.dev/handbook/web-dev-components/#glitches)). Please delete all of these `allow` attributes in your embedded Glitches unless you actually need these permissions. It seems like a potential attack surface that I'd like to avoid.

  </figcaption>
</figure>

What you can do is create two play-pending animations (`openModal`, and an inline opacity transformation), and then pause one of the animations, delaying it until the other is finished. You can then use promises to wait for each to be finished before playing. Finally, you can check to see if a flag is set, and then reverse each animation.

<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
  <iframe
    src="https://glitch.com/embed/#!/embed/waapi-promises?path=script.js&previewSize=0"
    title="waapi-promises on Glitch"
    allow="geolocation; microphone; camera; midi; vr; encrypted-media"
    style="height: 100%; width: 100%; border: 0;">
  </iframe>
</div>

### Example: Dynamic interactions with partial keyframes

<figure class="w-figure">
<video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/web-animations/retargetting.mp4">
  </video>  <figcaption class="w-figcaption">
    Retargeting example, where a mouse click adjusts the animation to a new location. <a href="https://codepen.io/una/details/BaoveQJ">See Demo on Codepen </a>
  </figcaption>
</figure>

```js

> selector.animate([{transform: `translate(${x}px, ${y}px)`}], {duration: 1000, fill: 'forwards'});
> }

Reviewer 2: ```suggestion
selector.animate([{transform: `translate(${x}px, ${y}px)`}], 
    {duration: 1000, fill: 'forwards'});
```
Removes extra `}` and avoids horizontal overflow 

![image](https://user-images.githubusercontent.com/4713486/83058750-b468af80-a00d-11ea-8ee4-0f54fccf4438.png)
